### DISCUSSION TITLE
Is it possible to build the LSP module with local dependencies?

### DISCUSSION THREAD

**Discussion by schuelermine** [orbitalquark/textadept#390] [Q&A]
I’m using NixOS, and I want to use the LSP module. NixOS best practice requires that all builds are reproducible, so I can’t fetch the prebuilt LSP module from GitHub, as it is not versioned. I want to build the LSP module declaratively but this usually has internet access disabled for reproducibility, expecting all dependencies to be downloaded in separate “fixed-output derivations” with hash specifiers. NixOS has Lua modules available under the `luaPackages` attribute set. But looking at the CMakeLists.txt it looks like the build process unconditionally downloads most dependencies. Can I specify a local path for these dependencies instead?

**Comment by mhwombat** [comment]
I wrote a NixOS package for Textadept 12.0, which is also built using CMake. It's a draft pull request because v12.0 hasn't officially been released yet, but I have tested it and it works. The basic approach is to download the files (as dependencies) before the build step. You can probably follow the same approach.

https://github.com/NixOS/nixpkgs/pull/214780

**Reply by schuelermine** [reply]
But looking at the CMakeLists.txt file, it looks like Textadept itself provides facilities for providing dependencies in ${deps_dir}. But the textadept-lsp CMakeLists.txt doesn’t! It seems to only check for dkjson.
```cmake
set(deps_dir ${CMAKE_BINARY_DIR}/_deps)
set(dkjson_tgz dkjson-2.5.tar.gz)
set(dkjson_url file://${deps_dir}/${dkjson_tgz})
if(NOT EXISTS ${deps_dir}/${dkjson_tgz})
  set(dkjson_url http://dkolf.de/src/dkjson-lua.fsl/tarball/${dkjson_tgz})
endif()
FetchContent_Declare(dkjson URL ${dkjson_url})
FetchContent_MakeAvailable(dkjson)
```
but other just seem to fetch unconditionally (I don’t know exactly how CMake works, so I might be wrong)
```cmake
set(penlight_zip 1.13.1.zip)
FetchContent_Declare(penlight
  URL https://github.com/lunarmodules/Penlight/archive/refs/tags/${penlight_zip})
FetchContent_MakeAvailable(penlight)
```
Can I override this somehow? penlight, lualogging and ldoc are all available on NixOS from `luaPackages`.

**Reply by schuelermine** [reply]
Or does putting them into `_deps` accomplish that?

**Reply by orbitalquark** [reply]
Yes, putting them in `_deps` tells CMake not to download them.

**Reply by schuelermine** [reply]
Even if there isn’t an explicit check for `NOT EXISTS ${deps_dir}/…`?

**Reply by orbitalquark** [reply]
No, that explicit check is needed. You'd want to do that for other archives too. Sorry for the confusion.

**Reply by schuelermine** [reply]
OK, I will add in a patch that checks for them.

**Comment by mhwombat** [comment]
Have you thought about how to make the LSP module available to Textadept? Textadept expects to find all modules, including LSP, in the `modules` folder in the user config directory. I'd like to have something like `emacsWithPackages` or `vimWithPackages`, but for Textadept. I haven't had time yet to work on it, or even think much about the best way to approach it. If you have any thoughts on the subject, let me know.

**Reply by schuelermine** [reply]
I'll be using home-manager to create a symlink to the built module in .textadept/modules/

**Comment by paaguti** [comment]
To add to the discussion...

A long time user of TextAdept, I'm encouraging my students to use it. I
don't know if this may help, but I'm building Debian packages for TextAdept
since ages and
after a minimal quirk has been fixed, I'm also building (and using)
textadept-lsp. This is what I currently build from the sources:

textadept-modules-file-diff_12.0nightly-4_amd64.deb
textadept-modules-lsp_12.0nightly-4_all.deb
textadept-modules-lua-repl_12.0nightly-4_all.deb
textadept-modules-yaml_12.0nightly-4_all.deb
textadept_12.0nightly-4_amd64.deb

Hope it helps, /PA

On Sun, 9 Apr 2023 at 01:49, Amy de Buitléir ***@***.***>
wrote:

> Have you thought about how to make the LSP module available to Textadept?
> Textadept expects to find all modules, including LSP, in the modules
> folder in the user config directory. I'd like to have something like
> emacsWithPackages or vimWithPackages, but for Textadept. I haven't had
> time yet to work on it, or even think much about the best way to approach
> it. If you have any thoughts on the subject, let me know.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/390#discussioncomment-5562001>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4BUKQDM7VMWCBGQBVTXAH2QLANCNFSM6AAAAAAWXXLKKI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>


-- 
Fragen sind nicht da, um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

Headaches with a Juju log:
unit-basic-16: 09:17:36 WARNING juju.worker.uniter.operation we should run
a leader-deposed hook here, but we can't yet
